src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -642,40 +678,27 @@ private static void PrepareReflectionObjects()
             s_connectionType = systemNetHttpAssembly?.GetType("System.Net.Connection");
             s_writeListField = s_connectionType?.GetField("m_WriteList", BindingFlags.Instance | BindingFlags.NonPublic);
 
-            // Second step: Generate an accessor for HttpWebRequest._HttpResponse
-            FieldInfo responseField = typeof(HttpWebRequest).GetField("_HttpResponse", BindingFlags.NonPublic | BindingFlags.Instance);
-            if (responseField != null)
-            {
-                string methodName = responseField.ReflectedType.FullName + ".get_" + responseField.Name;
-                DynamicMethod getterMethod = new DynamicMethod(methodName, typeof(HttpWebResponse), new Type[] { typeof(HttpWebRequest) }, true);
-                ILGenerator generator = getterMethod.GetILGenerator();
-                generator.Emit(OpCodes.Ldarg_0);
-                generator.Emit(OpCodes.Ldfld, responseField);
-                generator.Emit(OpCodes.Ret);
-                s_httpResponseAccessor = (Func<HttpWebRequest, HttpWebResponse>)getterMethod.CreateDelegate(typeof(Func<HttpWebRequest, HttpWebResponse>));
-            }
+            s_httpResponseAccessor = CreateFieldGetter<HttpWebRequest, HttpWebResponse>("_HttpResponse", BindingFlags.NonPublic | BindingFlags.Instance);
+            s_autoRedirectsAccessor = CreateFieldGetter<HttpWebRequest, int>("_AutoRedirects", BindingFlags.NonPublic | BindingFlags.Instance);
+            s_coreResponseAccessor = CreateFieldGetter<HttpWebRequest, object>("_CoreResponse", BindingFlags.NonPublic | BindingFlags.Instance);
 
-            // Third step: Generate an accessor for HttpWebRequest._AutoRedirects
-            FieldInfo redirectsField = typeof(HttpWebRequest).GetField("_AutoRedirects", BindingFlags.NonPublic | BindingFlags.Instance);
-            if (redirectsField != null)
+            s_coreResponseDataType = systemNetHttpAssembly?.GetType("System.Net.CoreResponseData");
+            if (s_coreResponseDataType != null)
             {
-                string methodName = redirectsField.ReflectedType.FullName + ".get_" + redirectsField.Name;
-                DynamicMethod getterMethod = new DynamicMethod(methodName, typeof(int), new Type[] { typeof(HttpWebRequest) }, true);
-                ILGenerator generator = getterMethod.GetILGenerator();
-                generator.Emit(OpCodes.Ldarg_0);
-                generator.Emit(OpCodes.Ldfld, redirectsField);
-                generator.Emit(OpCodes.Ret);
-                s_autoRedirectsAccessor = (Func<HttpWebRequest, int>)getterMethod.CreateDelegate(typeof(Func<HttpWebRequest, int>));
+                s_coreStatusCodeAccessor = CreateFieldGetter<HttpStatusCode>(s_coreResponseDataType, "m_StatusCode", BindingFlags.Public | BindingFlags.Instance);
+                s_coreHeadersAccessor = CreateFieldGetter<WebHeaderCollection>(s_coreResponseDataType, "m_ResponseHeaders", BindingFlags.Public | BindingFlags.Instance);
```

Comment:
Weird that these fields are public, since that you are using BindingFlags.Public.
I wonder if it's worthwhile to make it more robust by doing BindingFlags.Public | BindingFlags.NonPublic, so it would work if they later become private fields.

Reply:
>Weird that these fields are public

They are public fields but the type is internal.  So, it isn't really "public" except for internal use in the System.dll assembly.

Reply:
> so it would still work if they later become private fields.

I think we cannot protect this code from future changes in HTTP stack. If it changes, likely other places in this hook will break